### DISCUSSION TITLE
Booleans without Prefixes

### DISCUSSION THREAD

**Discussion by nateklaiber** [SPSCommerce/sps-api-standards#32] [Q&A] [answered]
I often include hints in resource objects. Things like `has_items` or `is_owner` etc. This is used as a hint and clear way to get an answer without having to directly look at associated resources (collection or instance).

The [documentation states](https://github.com/SPSCommerce/sps-api-standards/blob/main/standards/naming.md)

> Boolean properties SHOULD NOT use is, has, or another prefix.

And yet, if I want to relay those items as actual objects, then `items` or `owner` conflicts with those as objects. 

In fact, the documentation does not propose how to handle this. It shoes a noun as incorrect ('item'), but then the _correct_ has things like 'active', 'clever', and 'enabled'. 

I believe without the prefix the intent is not clear. What is the recommendation in situations above?

**Comment by ca0abinary** [comment]
@nateklaiber Does it make sense in your scenario to use these guidelines to improve the usefulness of the data communicated? For example: `has_items` is good information, but maybe something like `childItemCount` would convey not only that there are child items but how many there are enabling enhanced decision making without querying the child items. In the case of `is_owner` / `owner`, it's not clear to me what that means and how there would be two instances of the `owner` property. Maybe the property names could get away with some disambiguation?

**Comment by nateklaiber** [comment]
@ca0abinary - very good points. I tend to include them both in different scenarios. Counter caches are important. Query methods are important. 

I include counter caches where necessary (and move to a summary table if there are scoped counter caches).

I include query methods as helpful hints. A peek-ahead into the modeling without having to ask the modeling. I do this in server side code as well. I may have things like:

```
record.items?
record.hasItems()
record.items_count
record.itemsCount()
record.items
record.items()
```

> Various approaches for different languages and support.

In a server side language, asking `if record.items? then 'do something' else 'do something else' end` is preferred over `if record.items_count > 0 then 'do something' else 'do something else' end`

My goal in including both is to not bind them together, but also let them serve their purposes.

Here's another example for an _instance_. In this case I have a 'has one' relationship on my resource. Having something like:

```
document.has_asset
```
Makes more sense to me than:

```
document.asset_count
```
Which will always be 0 or 1.

Again, I treat this as a query method related to the various associations. I find representing them in the API with these prefixes is helpful.

**Reply by ca0abinary** [reply]
@nateklaiber Thanks for a great explanation! I wonder if you can "have your cake and eat it too" here by providing for the additional features on the client side.

For example:

Server API provides a single integer value:
```
record.itemCount
```

Where the client implementation expands on that (pseudo-code alert):
```
ClientRecord
{
    List<ServerItemRecord> _items = null;
    List<ServerItemRecord> items = ServerRecord.itemCount > 0 ? getItemsFromCache() : null;
    boolean hasItems = () => ServerRecord.itemCount > 0;
    int items_count = ServerRecord.itemCount;
    int itemsCount = () => ServerRecord.itemCount;
    ...
}
```

In this example all the convenience functions are implemented client-side (be that a server client or user client) and can key off of one server side count.

Hopefully I'm understanding your ask correctly. 😄

**Reply by travisgosselin** [reply]
Including convenience properties in the API contract is nice, and maybe completely a benefit in some cases. In other cases, it serves to make contract design a bit uglier over time as that contract evolves. More importantly, though, I think as @ca0abinary is showing, the convenience functions you are looking for are automatically available in most languages.... `serverRecord.Items.Any()` (C#), `serverRecord.items.isEmpty()` (Java), `_(serverRecord.items).isEmpty()` (JavaScript). Additionally, you could include these functions as first-class properties of your model to have them available in the client SDK / HTTP Wrapper.

**Reply by nateklaiber** [reply]
Thanks for the response @travisgosselin 👍 

I agree with you on contract design and it getting uglier. I think that's also a bigger and separate topic, as it has more to do with modeling the architecture at the gateway of your API as your systems evolve. This includes many other topics, too. 

I do understand they're available in other languages. I use them often. I mostly spend time writing in Ruby, Python, and JavaScript as my top 3.

I also like to address separation between SDKs and the general client.

I still keep things like this in my server modeling:

```ruby
def items?
  self.items.any?
end

def item?
  self.item.present?
end
```

This way I'm not conflating the concept of counter caches with presence in an instance or collection. Yes, they may be related, but I don't bind them as such and create the query method instead (or, in other languages I add the prefix to differentiate like `record.items()` and `record.hasItems()` - where one is a boolean check, and the other is the association).

I also often use `jq` on the command line, or another HTTP editor where I interact via `curl` or other tool. In that case it is easy to query where I can check a boolean response than the counter cache. In this case, there is no SDK around it yet. It's simply discovery or quick querying of the API.

Anyway, you've given me more food for thought, but my thought here is I still want to keep separation of concerns between what it means to 'have items' and 'items count' - and not bind them together (even though it may always be the case they are related in some fashion). 

Thanks for the discussion.

**Comment by travisgosselin** [comment, marked as answer]
@nateklaiber - great question. I too have often used these types of prefixes in the past in some APIs I have built. In putting together these standards with our working group, and looking for consistency of usage in our terminology, we found that usage of such prefixes often were shortcuts to more fundamental questions we should be asking about our API contract.

If wanting a field like "is_owner", it's often a bit ambiguous in terms of what this owner is (obviously context is important and difficult to talk about in examples without some). If not including the actual "owner" complex object, then including a property name in the contract that more fundamentally describes the owner is encouraged. This should take place with an adjective (something that describes it). You might include an adjective to the "owner" noun (an adjective modifier) to do that effectively: "active_owner" if needed. 

In terms of convenience properties for collections, the intent is that SDKs or language collection syntax is a preferred location to handle that.

**Reply by nateklaiber** [reply]
I wanted to apologize for my poor example of `is_owner` without giving you more context. I'd like to zoom in a bit on the example I provided with the has one relationship of a resource and associated asset (IO). I may be missing something so I'd love to get more insight and feedback on how you'd address or solve.

This is often the case in financial applications or others where you get to define a "receipt" resource (or "document"). You can define that without providing the actual file. You can later add that file (asset). 

> Note: This could be handled as multipart as well, but I still build these as resources where the asset is extended data but not always required.

If I am looking at the Resource of something like a Receipt, then I often want to know if it has a related asset. I can use that in advance of retrieving the separate resource request (`/receipts/{id}/asset`) where I may want it as a specific MIME type (PDF, PNG, etc). 

In this case, it is more of a query method helper than a counter cache. Yes, counter cache will answer if I have something, but the counter cache isn't used for anything. It's always 0 (no asset) or 1 (has asset).

And, in this case, part of my code for checking if there is an asset is ensuring we 1) have the underlying asset reference (polymorphic) [`self.count > 0` or other], and 2) ensuring we have the file in storage/persistence (Disk, memory, etc). The database record is present and we have the storage reference. Those things combined are what tell me if `record.asset?` is true, not just the count of the database record.

> Note: I check the existence in case a binding was broken between the DB and storage mechanism, or in case something became corrupt. Both of which will raise errors to me on the server side, but not break the front-end contract of the API/platform.

I don't believe these are SDK concerns just yet. I think they're still related to the intent and communications of the platform itself, before any SDK is built.

With these responses I also show corresponding hypermedia to those resources. For instance, if I have a file on disk, I provide a link to the REST resource of that object, as well as the permalink of the file from our asset server itself (optimized asset subdomain with precompiled assets and caching). BUT, this attribute is a helper method to receive in the initial response without making a separate request.

I also have more thoughts on the separation of concerns between the platform and SDK, but I will save that for another discussion topic.

Again, thanks so much for the feedback.

**Reply by ca0abinary** [reply]
@nateklaiber Thanks for the example! I'm wondering if in a 1:1 scenario if it might be helpful to have the asset id. If that's `null / undefined` (javascript) then there is no related asset (due to lack of asset, storage issues, etc.) otherwise the id indicates an intact asset can be used as part of a later request for that specific asset.

Where I'm a little lost is it seems like `is_owner` could indicate ownership of multiple assets without clearly defining which. Is that a correct understanding? If so, as a developer I would expect `receiptId` to be defined or not and operate on that reference (and any other asset types would have a specific identifier as well).

**Reply by travisgosselin** [reply]
@nateklaiber  If you have an Open API spec you wanted to share with the above-mentioned scenario, might make it easier to discuss/clarify?

**Reply by nateklaiber** [reply]
@ca0abinary @travisgosselin - I think it's important to find our common root for the discussion. I wanted to confirm that we're in agreement on most topics here. More on that shortly. Also, many things I'm referencing here are in private APIs that I'm not at liberty to share anything (But, I will work up some real world examples that we can all share and review).

I wanted to address the point of `asset_id`. For many of my resources I won't expose the ID for a has one relationship. You can only get it through `/receipts/{id}/asset`. I don't provide it as a standalone resource. So there is no `asset_id` or `asset_count` in these cases. However, we still want to be able to provide the client with the knowledge that there is indeed an asset at the other end (and it exists in storage without issues). This is by design. I don't want you to store or have reference to an ID in these cases. If you need to update the asset, you must do it through that endpoint, and the logic behind the scenes may handle paranoid deletes, versions, or time series. The point of entry is always through the parent resource. You also can't POST to something like `/receipts/{id}/assets` to create a new asset. It's a singular resource.

**Reply by nateklaiber** [reply]
As another example, the [SPS Commerce API Documentation for Shipping Labels](https://developercenter.spscommerce.com/#/docs/shipping-doc-api/labels/get_labels_by_id) has an attribute of `canRender`. What approach would you take to remove this prefix? Or is `can_` an acceptable prefix? Isn't this a boolean prefix? Without this prefix it would be `render`, correct? How would a developer interpret this?

**Reply by travisgosselin** [reply]
RE: Private APIs / Endpoints - sure, we have many internal APIs as well right now we don't expose. Sometimes when working in public forums we produce a smaller API spec as an example demonstration that is not exactly related to the original - but like you say, sometimes you can even lose context in these.

RE:`asset_id` - feels like you have structured this well based on the scenario from what I understand. If a resource does not require its own endpoint then don't expose it (don't expose anything you don't need too)... always not exposing an ID if its not required is a huge advantage as you evolve that API. This is the same spirit of reason for this line item:

> Resource identifiers SHOULD NOT be serialized as numbers or integers (of any precision) as they should be considered opaque values to the API consumer.

https://spscommerce.github.io/sps-api-standards/standards/serialization.html#number

RE: `canRender` - As you have already seen, our public APIs do not yet follow these API Standards. Our biggest transformations right now are happening internally. For this specific scenario of `canRender` I would consider joining the verb `render` with an adjective like "availableRendering" or "renderingAvailable"... or "render" might be the wrong term entirely to use here if there is something more specific or a type of rendering. Its worth noting as well that this item in the API Standards is a `SHOULD` leaving some room for context specific decisions on best terminology in the domain. If a team was really adamant on using "canRender", that may be just fine.

**Comment by JamesStauffer** [comment]
We have an API with a has_views property that is manually kept in sync with the DB to tell the code if it should even query for views. Comments?

**Reply by travisgosselin** [reply]
This sounds like the REST endpoint is formulated to represent a database table or object rather than modeling a resource. I'd start by investigating what the resource would be and decoupling it from database concepts. At that point, perhaps a more "adjective" friendly name will materialize to represent what the views are describing in this case.

**Reply by JamesStauffer** [reply]
I'm drawing a blank trying to think of a good adjective friendly name. "viewable" would be too confusing.
